cluster-autoscaler/cloudprovider/aws/api/instance_info.go

```diff
@@ -0,0 +1,342 @@
+					if price.EndRange != "Inf" || price.Unit != "Hrs" {
+						continue
+					}
+					p, err := strconv.ParseFloat(price.PricePerUnit.USD, 64)
+					if err != nil {
+						return fmt.Errorf("error parsing price for SKU %s [%s] %v", sku, price.PricePerUnit.USD, err)
+					}
+
+					if p == 0.0 {
+						continue
+					}
+
+					if lastOfferTime.IsZero() || lastOfferTime.After(offer.EffectiveDate) {
+						lastOfferTime = offer.EffectiveDate
+						lastOfferPrice = p
+					}
+				}
+			}
+
+			i.OnDemandPrice = lastOfferPrice
+		}
+
+		instances = append(instances, i)
+	}
+
+	bucket.Clear()
+	bucket.Add(instances...)
+	bucket.SetLastSync()
+
+	return nil
+}
+
+func (s *instanceInfoService) fetch(availabilityZone string, etag string) (*response, error) {
+	url := fmt.Sprintf(awsPricingAPIURLTemplate, availabilityZone)
+
+	req, err := http.NewRequest("GET", url, nil)
+
+	if len(etag) != 0 {
+		req.Header.Add("If-None-Match", etag)
+	}
+
+	res, err := s.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching [%s]", url)
+	}
+
+	defer res.Body.Close()
+
+	if res.StatusCode == 304 {
+		return nil, nil
+	}
+
+	var body []byte
+	if body, err = ioutil.ReadAll(res.Body); err != nil {
+		return nil, fmt.Errorf("error loading content of %s", url)
+	}
+
+	if res.StatusCode != 200 {
+		return nil, fmt.Errorf("got unexpected http status code %s with body [%s]", res.StatusCode, string(body))
+	}
+
+	var data = new(response)
+	if err := json.Unmarshal(body, data); err != nil {
+		return nil, fmt.Errorf("error unmarshaling %s with body [%s]", url, string(body))
+	}
+
+	return data, nil
+}
+
+type instanceInfoCache map[string]*regionalInstanceInfoBucket
+
+type regionalInstanceInfoBucket struct {
+	lastSync time.Time
+	ETag     string
+	mu       sync.RWMutex
+	info     []InstanceInfo
+}
+
+func (b *regionalInstanceInfoBucket) SetLastSync() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.lastSync = time.Now()
+}
+
+func (b *regionalInstanceInfoBucket) LastSync() time.Time {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	return b.lastSync
+}
+
+func (b *regionalInstanceInfoBucket) Clear() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.info = make([]InstanceInfo, 0)
+}
+
+func (b *regionalInstanceInfoBucket) Add(info ...InstanceInfo) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.info = append(b.info, info...)
+}
+
+type response struct {
+	Products map[string]product `json:"products"`
+	Terms    terms              `json:"terms"`
+}
+
+type terms struct {
+	OnDemand map[string]productOffers `json:"OnDemand"`
+}
+
+type productOffers map[string]productOffer
+
+type productOffer struct {
+	OfferTermCode   string                           `json:"offerTermCode"`
+	EffectiveDate   time.Time                        `json:"effectiveDate"`
+	SKU             string                           `json:"sku"`
+	PriceDimensions map[string]productPriceDimension `json:"priceDimensions"`
+}
+
+type productPriceDimension struct {
+	RateCode     string       `json:"rateCode"`
```

Comment:
either all privte or all public.

Reply:
Here and elsewhere.

Reply:
In my opinion, the struct `productPriceDimension` is just a local DTO used to represent the json structure. Therefore, it's not part of packages api and doesn't need to be kept stable/backward compatible. Unfortunately, the marshaller is not able to read private fields of foreign structs. 
This is the reason why I've decided to make it like this. Does that work for you?

Reply:
ok

Reply:
ok, if we are super sure that this cannot be done with the official library.